src/util/fs.cc

```diff
@@ -121,4 +127,377 @@ bool os_portable_copy_file_to_dir(const std::string& src,
   return ret == 0;
 }
 
+/// Extract path from path
+/// C:\a\b\c.txt -> "C:\a\b\" 
+/// C:\a\b\c -> C:\a\b
+/// d/e/ghi  -> d/e/
+std::string os_portable_path_from_path(const std::string& path) {
+
+  std::string sep;
+#if defined(_WIN32) || defined(_WIN64)
+  // on windows
+  sep = "\\";
+#else
+  // on *nix
+  sep = "/";
+#endif
+  if (path.find(sep) == path.npos)
+    return "./";
+  auto pos = path.rfind(sep);
+  ILA_ASSERT(pos != path.npos);
+  return path.substr(0,pos+1); // include sep
+}
+
+/// C:\\a.txt -> C:\\a   or  /a/b/c.txt -> a/b/c
+std::string os_portable_remove_file_name_extension(const std::string fname) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Function parameter 'fname' should be passed by const reference.](https://app.codacy.com/app/Bo-Yuan-Huang/ILAng/pullRequest?prid=4555809)